Docs: Extend guidelines for managing packages and publishing them to npm

gziolo opened the pull request:
## Description

As discussed during the [JavaScript weekly chat](https://make.wordpress.org/core/2018/08/07/javascript-chat-summary-august-7/) earlier this week, I opened pull request to extend our guidelines for managing and publishing npm packages with Lerna.

### TODO
* [x] creating new packages
* [x] Babel dependencies
* [x] maintaining changelogs

gziolo committed:
Docs: Extend guidelines for publishing packages to npm

gziolo committed:
Docs: Add details about creating new package

gziolo commented:
**Problem**: Publishing modules to npm can be difficult because it is not always clear at the time of publishing to what version the package should be updated. The individual performing the publish is not always the same as the contributors who had proposed changes to the package.

**Options**:

* Adopt conventional commits to enable automatic versioning (pull request)
  * Issue: This is hard to enforce
* Encourage developers to introduce relevant changelog entries as part of proposed changes. The type of change should indicate versioning requirements.
  * Issue: Not every version bump is caused by directly by new changes, sometimes indirectly by updated dependencies.
  * Issue: Prone to frequent merge conflicts with many developers adding entries to the same text file.

**Decision**: Encourage developers to introduce changelog entries and evaluate success in a future meeting.

gziolo commented:
@hypest feel free to include `react-native` in the `package.json` example. We probably should add a note that it should be added only for packages which are meant to be sent to the user. It doesn't make sense in the context of build tools.

hypest committed:
Add React Native entrypoint in example package.json

gziolo commented:
That’s a really good idea. We can almost copy and paste what you described. I would also encourage grouping by:
- Breaking
- New
- Fix
And skip everything else from the changelog. This is what popular repositories do usually.

aduth commented:
A correction to my previous comment, which is worth clarifying as it's a potential point of confusion:

>If I need to add something considered a bugfix, I add the "Fix:" item and change version to 1.2.3

In my above example, since the existing "1.2.2" version is marked unreleased, I shouldn't change the version at all.

Similarly, if the version was "1.3.0 (Unreleased)" and I was adding a new feature, I wouldn't change the version.

The version bump is only necessary if (a) there are no other unreleased changes or (b) the type of change I'm introducing is incompatible (more severe) than the other unreleased changes.

gziolo committed:
Docs: Add managing packages section

gziolo committed:
Docs: Tiny fix in CONTRIBUTING.md file

gziolo committed:
Docs: Tiny title fix in CONTRIBUTING.md file

gziolo commented:
This is ready for a final check. I included all suggestions shared by @aduth. Thanks for help 💯

ntwb committed:
Minor verbiage tweak to _Major version_ semver

ntwb committed:
Revert accidental changes added in 3a034a1

ntwb reviewed (approved):
LGTM

aduth commented on CONTRIBUTING.md:
````diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+		}
+	}
+	```
+	It assumes that code is goind to be located in `src` folder and will be transpiled with `Babel`.
````
Typo: "goind" -> "going"

aduth commented on CONTRIBUTING.md:
````diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+	```
+3. `README.md` file containing at least:
+	- Package name.
+	- Package description.
````
Same as in `package.json` ? (Should we mention this?)

gziolo replied:
It isn't always the case. Not sure to be honest.

aduth commented on CONTRIBUTING.md:
```diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+	- Package description.
+	- Installation details.
+	- Usage example.
+	- `Code is Poetry` logo.
```
Where does one find this?

gziolo replied:
We should include the snippet. It's copy and paste anyways.

aduth commented on CONTRIBUTING.md:
```diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+	
+### Maintaining changelogs
+
+It isn't easy task to maintain dozens of npm packages. That's why we decided to introduce `CHANGELOG.md` files for all packages to simplify the release process.
```
Grammar: "It isn't an easy"

aduth commented on CONTRIBUTING.md:
````diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+```md
+## v1.2.2 (Unreleased)
+
+#### Bug Fix
````
Based on document structure, why would this be an `H4` following an `H2` ?

gziolo replied:
I blame `Babel`, will update :)

aduth commented on CONTRIBUTING.md:
```diff
@@ -109,25 +109,109 @@ The workflow is documented in greater detail in the [repository management](./do
+
+- If you need to add something considered a bug fix, you add the item to `Bug Fix`section and leave the version as 1.2.2.
+- If it's a new feature you add the item to `New Feature` section and change version to 1.3.0.
+- If it's a breaking change you want to introduce, you add the item to `Breaking Change` section and bump the version to 2.0.0.
```
Are there any other changes which wouldn't fall neatly under one of these classifications? I've personally observed some "Internal" / "Housekeeping" type changes which don't necessarily contribute to public-facing consumer, though I suppose could have some place in the SemVer hierarchy. These are also items it's not entirely clear we'd need to include at all.

gziolo replied:
> These are also items it's not entirely clear we'd need to include at all. 

It might get tricky because Lerna will bump version anyways. We probably should skip them as they aren't that important but we still will have a new version released. It might sometimes happen only because one of the packages maintained by Lerna listed in dependencies has version bump.

tofumatt committed:
docs: Make `npm` consistent and tweak a few things

tofumatt committed:
docs: Tweaks for readability, typo fixes, etc.

tofumatt reviewed (approved):
Sorry this took me ages to get to, I got distracted with some other PRs! 😓

I made a few tweaks and tidied up some of the docs based on @aduth's comments. The rest looks good to me. 👍

aduth commented on CONTRIBUTING.md:
```diff
@@ -109,45 +109,130 @@ The workflow is documented in greater detail in the [repository management](./do
+	
+### Maintaining changelogs
+
+Maintaining dozens of npm packages is difficult–it can be tough to keep track of changes. That's why we use `CHANGELOG.md` files for each package to simplify the release process. All packages should follow the [Semantic Versioning (`semver`) specification](https://semver.org/).
```
The ultimate nitpick! If we're going to be fancy with dashes, we should use the correct one: here an [en dash](http://www.thepunctuationguide.com/en-dash.html) should be the [em dash](http://www.thepunctuationguide.com/em-dash.html) 😄 

Before: `–`
After: `—`

tofumatt replied:
Omg you’re totally right and I’m often the one to comment on that. I have no excuse!

tofumatt committed:
Use

tofumatt commented:
Sorry about my rubbish commit message, I made it from my iPad and fat-fingered the return key.

gziolo committed:
Update CONTRIBUTING.md

Added @aduth suggestion from the core JS chat: https://make.wordpress.org/core/2018/08/21/javascript-chat-summary-august-21/.

gziolo commented:
Thanks @tofumatt and @aduth for your help 🙇 Let's see how those guidelines influence daily workflow and iterate later.
